snapshot/snapshotfs/upload.go

```diff
@@ -423,117 +435,65 @@ func (u *Uploader) prepareWorkItems(ctx context.Context, dirRelativePath string,
-					},
-				})
-
-			case fs.File:
-				u.stats.NonCachedFiles++
-				result = append(result, &uploadWorkItem{
-					entry:             entry,
-					entryRelativePath: entryRelativePath,
-					uploadFunc: func() entryResult {
-						return u.uploadFileInternal(ctx, filepath.Join(dirRelativePath, entry.Name()), entry, policyTree.Child(entry.Name()).EffectivePolicy())
-					},
-				})
-
-			default:
-				return errors.Errorf("file type not supported: %v", entry.Mode())
-			}
+			results[index] = cachedDirEntry
+			return nil
 		}
-		return nil
-	})
-
-	return result, resultErr
-}
-
-func toChannel(items []*uploadWorkItem) <-chan *uploadWorkItem {
-	ch := make(chan *uploadWorkItem)
 
-	go func() {
-		defer close(ch)
+		switch entry := entry.(type) {
+		case fs.Symlink:
+			de, err := u.uploadSymlinkInternal(ctx, filepath.Join(dirRelativePath, entry.Name()), entry)
+			if err != nil {
+				return u.maybeIgnoreFileReadError(err, policyTree)
+			}
 
-		for _, wi := range items {
-			ch <- wi
-		}
-	}()
+			results[index] = de
 
-	return ch
-}
+		case fs.File:
+			atomic.AddInt32(&u.stats.NonCachedFiles, 1)
+			de, err := u.uploadFileInternal(ctx, filepath.Join(dirRelativePath, entry.Name()), entry, policyTree.Child(entry.Name()).EffectivePolicy())
+			if err != nil {
+				return u.maybeIgnoreFileReadError(err, policyTree)
+			}
 
-func (u *Uploader) launchWorkItems(workItems []*uploadWorkItem, wg *sync.WaitGroup) {
-	// allocate result channel for each work item.
-	for _, it := range workItems {
-		it.resultChan = make(chan entryResult, 1)
-	}
+			results[index] = de
 
-	workerCount := u.ParallelUploads
-	if workerCount == 0 {
-		workerCount = runtime.NumCPU()
-	}
+		default:
+			return errors.Errorf("file type not supported: %v", entry.Mode())
+		}
 
-	ch := toChannel(workItems)
+		return nil
+	})
 
-	for i := 0; i < workerCount; i++ {
-		wg.Add(1)
+	for _, ent := range results {
+		if ent == nil {
+			continue
+		}
 
-		go func() {
-			defer wg.Done()
+		if ent.Type == snapshot.EntryTypeFile {
+			u.stats.TotalFileCount++
+			u.stats.TotalFileSize += ent.FileSize
+			summ.TotalFileCount++
+			summ.TotalFileSize += ent.FileSize
 
-			for it := range ch {
-				it.resultChan <- it.uploadFunc()
+			if ent.ModTime.After(summ.MaxModTime) {
+				summ.MaxModTime = ent.ModTime
 			}
-		}()
-	}
-}
-
-func (u *Uploader) processUploadWorkItems(ctx context.Context, workItems []*uploadWorkItem, dirManifest *snapshot.DirManifest, ignoreFileErrs bool) error {
-	var wg sync.WaitGroup
-
-	u.launchWorkItems(workItems, &wg)
-
-	// Read result channels in order.
-	for _, it := range workItems {
-		result := <-it.resultChan
-
-		if result.err == errCancelled {
-			return errCancelled
 		}
+	}
 
-		if result.err != nil {
-			if ignoreFileErrs {
-				u.stats.ReadErrors++
-
-				log(ctx).Warningf("unable to hash file %q: %s, ignoring", it.entryRelativePath, result.err)
+	return notNilEntries(results), resultErr
```

Comment:
```suggestion
	return results, resultErr
```

Reply:
rewrote using channels, which is much cleaner